resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+  against the type on some path in some way.
+
+- `assigned` is a boolean value indicating whether the variable is known to
+  have been definitely assigned at the given point in the source code.
+
+- `unassigned` is a boolean value indicating whether the variable is known not
+  to have been definitely assigned at the given point in the source code.  (Note
+  that a variable may not be both definitely assigned and definitely
+  unassigned).
+
+- `writeCaptured` is a boolean value indicating whether a closure might exist at
+  the given point in the source code, which could potentially write to the
+  variable.
+
+A *flow model*, denoted `FlowModel(reachable, variableInfo)`, represents what
+is statically known to flow analysis about the state of the program at a given
+point in the source code.
+
+  - `reachable` is a stack of boolean values modeling the reachability of the
+  given point in the source code.  The `i`th element of the stack (counting from
+  the top of the stack) encodes whether the `i-1`th enclosing control flow split
+  is reachable from the `ith` enclosing control flow split (or when `i` is 0,
+  whether the current program point is reachable from the enclosing control flow
+  split).  So if the bottom element of `reachable` is `false`, then the current
+  program point is definitively known by flow analysis to be unreachable from
+  the first enclosing control flow split.  If it is `true`, then the analysis
+  cannot eliminate the possibility that the given point may be reached by some
+  path.  Each other element of the stack models the same property, starting from
+  some control flow split between the start of the program and the current node,
+  and treating the entry to the program as the initial control flow split.  The
+  true reachability of the current program point then is the conjunction of the
+  elements of the `reachable` stack, since each element of the stack models the
+  reachability of one control flow split from its enclosing control flow split.
+
+  - `variableInfo` is a mapping from variables in scope at the given point to
+  their associated *variable model*s.
+
+The following functions associate flow models to nodes:
+
+- `before(N)`, where `N` is a statement or expression, represents the *flow
+  model* just prior to execution of `N`.
+
+- `after(N)`, where `N` is a statement or expression, represents the *flow
+  model* just after execution of `N`, assuming that `N` completes normally
+  (i.e. does not throw an exception or perform a jump such as `return`, `break`,
+  or `continue`).
+
+- `true(E)`, where `E` is an expression, represents the *flow model* just after
+  execution of `E`, assuming that `E` completes normally and evaluates to `true`.
+
+- `false(E)`, where `E` is an expression, represents the *flow model* just after
+  execution of `E`, assuming that `E` completes normally and evaluates to `false`.
+
+- `null(E)`, where `E` is an expression, represents the *flow model* just after
```

Comment:
It's hard to say without knowing what the special behaviors around null are.  A less contrived example of something that I think this enables is the following:

```dart
foo(int? x, int? y) {
  if (x != null) {
    if (x == y) {
       y.isEven; // (1)
    }
  }
}
```

The line at `(1)` is allowed in my model because the true model for the comparison contains the join of the cases where either `x` and `y` are both null (which is empty since `x` is known not to be null) or that they are both non-null.  So `y` gets promoted in the continuation of the comparison.

Is this worth it?  I don't really know.  I didn't specify it this way specifically to enable this example, but it seemed to be the uniform way to do it.

Reply:
Seeing your comment later, I see that I have mis-specified operator ==, so I'll have to think further about the mechanism.  The example code still holds as something that I think I intended to work that I'm not sure works without tracking notNull.